### DISCUSSION TITLE
How can I trace the YOLO model training process step by step (file-by-file execution flow)?

### DISCUSSION THREAD

**Discussion by ManpreeetSingh29** [ultralytics/ultralytics#21036] [Q&A]
Hello team,
I want to understand the training process of a YOLO model in detail — specifically, which Python files and functions are executed sequentially from the moment I run the model.train()... command.
My goal is to dive into the backend and trace the flow step by step to better understand or possibly modify the training pipeline. For example:

- Which file gets called first when model.train() is run?
- How does it internally handle config parsing, dataset loading, model creation, and training loops?
- Is there a recommended way to log or trace the function calls for educational or debugging purposes?

I’ve looked through the repo but would love a high-level flow or file reference if one exists — or guidance on how best to approach this exploration.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ManpreeetSingh29, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response 🤖. An Ultralytics engineer will review your question and provide additional assistance soon.

**Reply by glenn-jocher** [reply]
Great question! Here's the step-by-step execution flow when you call `model.train()`:

1. **Entry Point**: [`ultralytics/engine/model.py`](https://docs.ultralytics.com/reference/engine/model/) - The `train()` method is called, which handles configuration parsing, creates the trainer instance, and manages HUB integration.

2. **Trainer Selection**: Based on your model task, it instantiates the appropriate trainer (e.g., [`DetectionTrainer`](https://docs.ultralytics.com/reference/models/yolo/detect/train/) for object detection from `ultralytics/models/yolo/detect/train.py`).

3. **Base Training Logic**: [`ultralytics/engine/trainer.py`](https://docs.ultralytics.com/reference/engine/trainer/) - The `BaseTrainer` class contains the core training loop with methods like dataset building, dataloader creation, model setup, and the main training iterations.

4. **Key Flow**: `model.train()` → `trainer.train()` → `BaseTrainer` methods (`build_dataset()`, `get_dataloader()`, `preprocess_batch()`, `set_model_attributes()`, etc.)

For tracing function calls, you can add logging or use Python's built-in `trace` module. The [Reference section](https://docs.ultralytics.com/reference/cfg/__init__/) of our docs provides detailed information about each file's functions and classes, which is perfect for understanding the internal architecture.
